test/helpers/utils.go

```diff
@@ -430,9 +430,12 @@ func CanRunK8sVersion(ciliumVersion, k8sVersionStr string) (bool, error) {
 // failIfContainsBadLogMsg makes a test case to fail if any message from
 // given log messages contains an entry from badLogMessages (map key) AND
 // does not contain ignore messages (map value).
-func failIfContainsBadLogMsg(logs string) {
+func failIfContainsBadLogMsg(logs string, whitelist map[string]bool) {
```

Comment:
Doesn't this replicate what the values in the badLogMessages does already? I would prefer to pass in the `badLogMessages` instead of added a slightly different whitelist mechanism. On the other hand, I'm not sure if that system is as flexible or simple as you need it since all you wanted here was to suppress blacklisting one message :/ (this is where I'm on the fence. Your new whitelist is a lot more clear)

Reply:
I wasn't sure which way to go either. If we pass `badLogMessages`, we would likely still want two `ValidateNoErrorsInLogs{,Default}` functions to avoid changing all call sites, but we wouldn't need to check the whitelist at each loop iteration.

However, if we pass `badLogMessages`, we'll need to change the default map for `K8sUpdates`. If we create a new map specifically for `K8sUpdates`, we risk missing future additions to `badLogMessages`. From my (still limited) understanding of Go, we would need to make a deep copy of `badLogMessages` to remove one entry (correct?), which sounds more involved than using a whitelist mechanism.

Reply:
While whitelist may seem clear, I think we should not change how this func works, and explicitly pass badLogMessages as an argument, as Ray suggested.

I think we should add a function which will copy global `badLogMessages` and return the copy, this way we make sure that we always work with fresh copy of the map, which can be manipulated per test suite.

As of this change, this function effectively gets two maps and calculates a difference between them, I think it should be on the caller to define "bad" log messages, and adding a whitelist adds unnecessary complexity into the function and makes it less readable.

We shouldn't have used a global variable in the first place, but adding whitelist seems like a hacky workaround instead of making the code better structured.

Reply:
yup, that is exactly what went through my mind too. You're also correct about the map deep-copy (unless they made `copy` do something in recent versions).
I guess my thinking came down to two questions (I'm not as worried as you about changing all the callsites despite it making the PR look much bigger):
1- what will people copy-paste?
2- will we ever debug a test because we think the whitelist is only the parameter, but then the built-in one is also taking effect.

In any case, you're point about missing updates to the original badLogMessages wins out, I think. I'll approve :)

Reply:
> I think we should add a function which will copy global `badLogMessages` and return the copy, this way we make sure that we always work with fresh copy of the map, which can be manipulated per test suite.

Makes sense. I'll push an updated branch in a bit.

Reply:
I've updated the branch with:
- A `GetBadLogMessages()` function to get a copy of the `badLogMessages`.
- A `ValidateListOfErrorsInLogs` function to pass a custom list of messages. `ValidateNoErrorsInLogs` calls `ValidateListOfErrorsInLogs` with the default `badLogMessages`.